File: flow/consumer-complaints.js
// To parse this data:
//
//   const Convert = require("./file");
//
//   const consumerComplaints = Convert.toConsumerComplaints(json);
//
// These functions will throw an error if the JSON doesn't
// match the expected interface, even if the JSON is valid.

// Converts JSON strings to/from your types
// and asserts the results of JSON.parse at runtime
function toConsumerComplaints(json) {
    return cast(JSON.parse(json), a(r("ConsumerComplaints")));
}

function consumerComplaintsToJson(value) {
    return JSON.stringify(uncast(value, a(r("ConsumerComplaints"))), null, 2);
}

function invalidValue(typ, val) {
    throw Error(`Invalid value ${JSON.stringify(val)} for type ${JSON.stringify(typ)}`);
}

function jsonToJSProps(typ) {
    if (typ.jsonToJS === undefined) {
        var map = {};
        typ.props.forEach((p) => map[p.json] = { key: p.js, typ: p.typ });
        typ.jsonToJS = map;
    }
    return typ.jsonToJS;
}

function jsToJSONProps(typ) {
    if (typ.jsToJSON === undefined) {
        var map = {};
        typ.props.forEach((p) => map[p.js] = { key: p.json, typ: p.typ });
        typ.jsToJSON = map;
    }
    return typ.jsToJSON;
}

function transform(val, typ, getProps) {
    function transformPrimitive(typ, val) {
        if (typeof typ === typeof val) return val;
        return invalidValue(typ, val);
    }

    function transformUnion(typs, val) {
        // val must validate against one typ in typs
        var l = typs.length;
        for (var i = 0; i < l; i++) {
            var typ = typs[i];
            try {
                return transform(val, typ, getProps);
            } catch (_) {}
        }
        return invalidValue(typs, val);
    }

    function transformEnum(cases, val) {
        if (cases.indexOf(val) !== -1) return val;
        return invalidValue(cases, val);
    }

    function transformArray(typ, val) {
        // val must be an array with no invalid elements
        if (!Array.isArray(val)) return invalidValue("array", val);
        return val.map(el => transform(el, typ, getProps));
    }

    function transformObject(props, additional, val) {
        if (val === null || typeof val !== "object" || Array.isArray(val)) {
            return invalidValue("object", val);
        }
        var result = {};
        Object.getOwnPropertyNames(props).forEach(key => {
            const prop = props[key];
            const v = Object.prototype.hasOwnProperty.call(val, key) ? val[key] : undefined;
            result[prop.key] = transform(v, prop.typ, getProps);
        });
        Object.getOwnPropertyNames(val).forEach(key => {
            if (!Object.prototype.hasOwnProperty.call(props, key)) {
                result[key] = transform(val[key], additional, getProps);
            }
        });
        return result;
    }

    if (typ === "any") return val;
    if (typ === null) {
        if (val === null) return val;
        return invalidValue(typ, val);
    }
    if (typ === false) return invalidValue(typ, val);
    while (typeof typ === "object" && typ.ref !== undefined) {
        typ = typeMap[typ.ref];
    }
    if (Array.isArray(typ)) return transformEnum(typ, val);
    if (typeof typ === "object") {
        return typ.hasOwnProperty("unionMembers") ? transformUnion(typ.unionMembers, val)
            : typ.hasOwnProperty("arrayItems")    ? transformArray(typ.arrayItems, val)
            : typ.hasOwnProperty("props")         ? transformObject(getProps(typ), typ.additional, val)
            : invalidValue(typ, val);
    }
    return transformPrimitive(typ, val);
}

function cast(val, typ) {
    return transform(val, typ, jsonToJSProps);
}

function uncast(val, typ) {
    return transform(val, typ, jsToJSONProps);
}

function a(typ) {
    return { arrayItems: typ };
}

function u(...typs) {
    return { unionMembers: typs };
}

function o(props, additional) {
    return { props, additional };
}

function m(additional) {
    return { props: [], additional };
}

function r(name) {
    return { ref: name };
}

const typeMap = {
    "ConsumerComplaints": o([
        { json: "id", js: "id", typ: "" },
        { json: "name", js: "name", typ: "" },
        { json: "averageRating", js: "averageRating", typ: 0 },
        { json: "createdAt", js: "createdAt", typ: 0 },
        { json: "description", js: "description", typ: "" },
        { json: "displayType", js: "displayType", typ: r("DisplayType") },
        { json: "downloadCount", js: "downloadCount", typ: 0 },
        { json: "hideFromCatalog", js: "hideFromCatalog", typ: true },
        { json: "hideFromDataJson", js: "hideFromDataJson", typ: true },
        { json: "indexUpdatedAt", js: "indexUpdatedAt", typ: u(undefined, 0) },
        { json: "newBackend", js: "newBackend", typ: true },
        { json: "numberOfComments", js: "numberOfComments", typ: 0 },
        { json: "oid", js: "oid", typ: 0 },
        { json: "provenance", js: "provenance", typ: r("Provenance") },
        { json: "publicationAppendEnabled", js: "publicationAppendEnabled", typ: true },
        { json: "publicationDate", js: "publicationDate", typ: 0 },
        { json: "publicationGroup", js: "publicationGroup", typ: 0 },
        { json: "publicationStage", js: "publicationStage", typ: r("PublicationStage") },
        { json: "rowClass", js: "rowClass", typ: u(undefined, "") },
        { json: "rowsUpdatedAt", js: "rowsUpdatedAt", typ: 0 },
        { json: "rowsUpdatedBy", js: "rowsUpdatedBy", typ: r("RowsUpdatedBy") },
        { json: "tableId", js: "tableId", typ: 0 },
        { json: "totalTimesRated", js: "totalTimesRated", typ: 0 },
        { json: "viewCount", js: "viewCount", typ: 0 },
        { json: "viewLastModified", js: "viewLastModified", typ: 0 },
        { json: "viewType", js: "viewType", typ: r("ViewType") },
        { json: "grants", js: "grants", typ: a(r("Grant")) },
        { json: "metadata", js: "metadata", typ: r("Metadata") },
        { json: "owner", js: "owner", typ: r("Owner") },
        { json: "ratings", js: "ratings", typ: u(undefined, r("Ratings")) },
        { json: "rights", js: "rights", typ: a(r("Right")) },
        { json: "tableAuthor", js: "tableAuthor", typ: r("TableAuthor") },
        { json: "flags", js: "flags", typ: u(undefined, a(r("ConsumerComplaintFlag"))) },
        { json: "moderationStatus", js: "moderationStatus", typ: u(undefined, true) },
        { json: "category", js: "category", typ: u(undefined, "") },
        { json: "tags", js: "tags", typ: u(undefined, a("")) },
        { json: "modifyingViewUid", js: "modifyingViewUid", typ: u(undefined, r("ModifyingViewUid")) },
    ], false),
    "Grant": o([
        { json: "inherited", js: "inherited", typ: true },
        { json: "type", js: "type", typ: r("GrantType") },
        { json: "flags", js: "flags", typ: a(r("GrantFlag")) },
    ], false),
    "Metadata": o([
        { json: "jsonQuery", js: "jsonQuery", typ: u(undefined, r("JSONQuery")) },
        { json: "rdfSubject", js: "rdfSubject", typ: u(undefined, "") },
        { json: "rdfClass", js: "rdfClass", typ: u(undefined, "") },
        { json: "rowIdentifier", js: "rowIdentifier", typ: u(undefined, "") },
        { json: "availableDisplayTypes", js: "availableDisplayTypes", typ: a(r("AvailableDisplayType")) },
        { json: "rowLabel", js: "rowLabel", typ: u(undefined, "") },
        { json: "renderTypeConfig", js: "renderTypeConfig", typ: r("RenderTypeConfig") },
        { json: "richRendererConfigs", js: "richRendererConfigs", typ: u(undefined, r("RichRendererConfigs")) },
        { json: "custom_fields", js: "custom_fields", typ: u(undefined, r("CustomFields")) },
    ], false),
    "CustomFields": o([
        { json: "TEST", js: "TEST", typ: r("Test") },
    ], false),
    "Test": o([
        { json: "CFPB1", js: "CFPB1", typ: "" },
    ], false),
    "JSONQuery": o([
        { json: "order", js: "order", typ: a(r("Order")) },
    ], false),
    "Order": o([
        { json: "ascending", js: "ascending", typ: true },
        { json: "columnFieldName", js: "columnFieldName", typ: "" },
    ], false),
    "RenderTypeConfig": o([
        { json: "visible", js: "visible", typ: r("Visible") },
    ], false),
    "Visible": o([
        { json: "table", js: "table", typ: u(undefined, true) },
        { json: "fatrow", js: "fatrow", typ: u(undefined, true) },
    ], false),
    "RichRendererConfigs": o([
        { json: "fatRow", js: "fatRow", typ: r("FatRow") },
    ], false),
    "FatRow": o([
        { json: "columns", js: "columns", typ: a(r("Column")) },
    ], false),
    "Column": o([
        { json: "styles", js: "styles", typ: r("Styles") },
        { json: "rows", js: "rows", typ: a(r("Row")) },
    ], false),
    "Row": o([
        { json: "fields", js: "fields", typ: a(r("Field")) },
    ], false),
    "Field": o([
        { json: "tableColumnId", js: "tableColumnId", typ: 0 },
        { json: "type", js: "type", typ: r("FieldType") },
    ], false),
    "Styles": o([
        { json: "width", js: "width", typ: r("Width") },
    ], false),
    "Owner": o([
        { json: "id", js: "id", typ: "" },
        { json: "displayName", js: "displayName", typ: "" },
        { json: "screenName", js: "screenName", typ: "" },
        { json: "type", js: "type", typ: r("OwnerType") },
        { json: "profileImageUrlLarge", js: "profileImageUrlLarge", typ: u(undefined, "") },
        { json: "profileImageUrlMedium", js: "profileImageUrlMedium", typ: u(undefined, "") },
        { json: "profileImageUrlSmall", js: "profileImageUrlSmall", typ: u(undefined, "") },
    ], false),
    "Ratings": o([
        { json: "rating", js: "rating", typ: 0 },
    ], false),
    "TableAuthor": o([
        { json: "id", js: "id", typ: r("ID") },
        { json: "displayName", js: "displayName", typ: r("Name") },
        { json: "screenName", js: "screenName", typ: r("Name") },
        { json: "type", js: "type", typ: r("OwnerType") },
    ], false),
    "DisplayType": [
        "fatrow",
        "table",
    ],
    "ConsumerComplaintFlag": [
        "default",
        "restorable",
        "restorePossibleForType",
    ],
    "GrantFlag": [
        "public",
    ],
    "GrantType": [
        "viewer",
    ],
    "AvailableDisplayType": [
        "fatrow",
        "page",
        "table",
    ],
    "FieldType": [
        "columnData",
        "columnLabel",
    ],
    "Width": [
        "27%",
        "30%",
        "31%",
        "33%",
        "40%",
    ],
    "ModifyingViewUid": [
        "s6ew-h6mp",
    ],
    "OwnerType": [
        "interactive",
    ],
    "Provenance": [
        "official",
    ],
    "PublicationStage": [
        "published",
    ],
    "Right": [
        "read",
    ],
    "RowsUpdatedBy": [
        "pjxg-ve4m",
        "54a3-qyun",
        "9e3m-2843",
        "vvca-fr6g",
    ],
    "Name": [
        "Dan Munz",
        "Doug Taylor",
        "ming",
        "Shannon Files",
    ],
    "ID": [
        "pjxg-ve4m",
        "54a3-qyun",
        "9e3m-2843",
        "9eki-h2hn",
    ],
    "ViewType": [
        "tabular",
    ],
};

module.exports = {
    "consumerComplaintsToJson": consumerComplaintsToJson,
    "toConsumerComplaints": toConsumerComplaints,
};

File: rustlang/consumer-complaints.rs
// Example code that deserializes and serializes the model.
// extern crate serde;
// #[macro_use]
// extern crate serde_derive;
// extern crate serde_json;
//
// use generated_module::[object Object];
//
// fn main() {
//     let json = r#"{"answer": 42}"#;
//     let model: [object Object] = serde_json::from_str(&json).unwrap();
// }

extern crate serde_json;

pub type ConsumerComplaints = Vec<ConsumerComplaint>;

#[derive(Serialize, Deserialize)]
pub struct ConsumerComplaint {
    #[serde(rename = "id")]
    id: String,

    #[serde(rename = "name")]
    name: String,

    #[serde(rename = "averageRating")]
    average_rating: i64,

    #[serde(rename = "createdAt")]
    created_at: i64,

    #[serde(rename = "description")]
    description: String,

    #[serde(rename = "displayType")]
    display_type: DisplayType,

    #[serde(rename = "downloadCount")]
    download_count: i64,

    #[serde(rename = "hideFromCatalog")]
    hide_from_catalog: bool,

    #[serde(rename = "hideFromDataJson")]
    hide_from_data_json: bool,

    #[serde(rename = "indexUpdatedAt")]
    index_updated_at: Option<i64>,

    #[serde(rename = "newBackend")]
    new_backend: bool,

    #[serde(rename = "numberOfComments")]
    number_of_comments: i64,

    #[serde(rename = "oid")]
    oid: i64,

    #[serde(rename = "provenance")]
    provenance: Provenance,

    #[serde(rename = "publicationAppendEnabled")]
    publication_append_enabled: bool,

    #[serde(rename = "publicationDate")]
    publication_date: i64,

    #[serde(rename = "publicationGroup")]
    publication_group: i64,

    #[serde(rename = "publicationStage")]
    publication_stage: PublicationStage,

    #[serde(rename = "rowClass")]
    row_class: Option<String>,

    #[serde(rename = "rowsUpdatedAt")]
    rows_updated_at: i64,

    #[serde(rename = "rowsUpdatedBy")]
    rows_updated_by: RowsUpdatedBy,

    #[serde(rename = "tableId")]
    table_id: i64,

    #[serde(rename = "totalTimesRated")]
    total_times_rated: i64,

    #[serde(rename = "viewCount")]
    view_count: i64,

    #[serde(rename = "viewLastModified")]
    view_last_modified: i64,

    #[serde(rename = "viewType")]
    view_type: ViewType,

    #[serde(rename = "grants")]
    grants: Vec<Grant>,

    #[serde(rename = "metadata")]
    metadata: Metadata,

    #[serde(rename = "owner")]
    owner: Owner,

    #[serde(rename = "ratings")]
    ratings: Option<Ratings>,

    #[serde(rename = "rights")]
    rights: Vec<Right>,

    #[serde(rename = "tableAuthor")]
    table_author: TableAuthor,

    #[serde(rename = "flags")]
    flags: Option<Vec<ConsumerComplaintFlag>>,

    #[serde(rename = "moderationStatus")]
    moderation_status: Option<bool>,

    #[serde(rename = "category")]
    category: Option<String>,

    #[serde(rename = "tags")]
    tags: Option<Vec<String>>,

    #[serde(rename = "modifyingViewUid")]
    modifying_view_uid: Option<ModifyingViewUid>,
}

#[derive(Serialize, Deserialize)]
pub struct Grant {
    #[serde(rename = "inherited")]
    inherited: bool,

    #[serde(rename = "type")]
    grant_type: GrantType,

    #[serde(rename = "flags")]
    flags: Vec<GrantFlag>,
}

#[derive(Serialize, Deserialize)]
pub struct Metadata {
    #[serde(rename = "jsonQuery")]
    json_query: Option<JsonQuery>,

    #[serde(rename = "rdfSubject")]
    rdf_subject: Option<String>,

    #[serde(rename = "rdfClass")]
    rdf_class: Option<String>,

    #[serde(rename = "rowIdentifier")]
    row_identifier: Option<String>,

    #[serde(rename = "availableDisplayTypes")]
    available_display_types: Vec<AvailableDisplayType>,

    #[serde(rename = "rowLabel")]
    row_label: Option<String>,

    #[serde(rename = "renderTypeConfig")]
    render_type_config: RenderTypeConfig,

    #[serde(rename = "richRendererConfigs")]
    rich_renderer_configs: Option<RichRendererConfigs>,

    #[serde(rename = "custom_fields")]
    custom_fields: Option<CustomFields>,
}

#[derive(Serialize, Deserialize)]
pub struct CustomFields {
    #[serde(rename = "TEST")]
    test: Test,
}

#[derive(Serialize, Deserialize)]
pub struct Test {
    #[serde(rename = "CFPB1")]
    cfpb1: String,
}

#[derive(Serialize, Deserialize)]
pub struct JsonQuery {
    #[serde(rename = "order")]
    order: Vec<Order>,
}

#[derive(Serialize, Deserialize)]
pub struct Order {
    #[serde(rename = "ascending")]
    ascending: bool,

    #[serde(rename = "columnFieldName")]
    column_field_name: String,
}

#[derive(Serialize, Deserialize)]
pub struct RenderTypeConfig {
    #[serde(rename = "visible")]
    visible: Visible,
}

#[derive(Serialize, Deserialize)]
pub struct Visible {
    #[serde(rename = "table")]
    table: Option<bool>,

    #[serde(rename = "fatrow")]
    fatrow: Option<bool>,
}

#[derive(Serialize, Deserialize)]
pub struct RichRendererConfigs {
    #[serde(rename = "fatRow")]
    fat_row: FatRow,
}

#[derive(Serialize, Deserialize)]
pub struct FatRow {
    #[serde(rename = "columns")]
    columns: Vec<Column>,
}

#[derive(Serialize, Deserialize)]
pub struct Column {
    #[serde(rename = "styles")]
    styles: Styles,

    #[serde(rename = "rows")]
    rows: Vec<Row>,
}

#[derive(Serialize, Deserialize)]
pub struct Row {
    #[serde(rename = "fields")]
    fields: Vec<Field>,
}

#[derive(Serialize, Deserialize)]
pub struct Field {
    #[serde(rename = "tableColumnId")]
    table_column_id: i64,

    #[serde(rename = "type")]
    field_type: FieldType,
}

#[derive(Serialize, Deserialize)]
pub struct Styles {
    #[serde(rename = "width")]
    width: Width,
}

#[derive(Serialize, Deserialize)]
pub struct Owner {
    #[serde(rename = "id")]
    id: String,

    #[serde(rename = "displayName")]
    display_name: String,

    #[serde(rename = "screenName")]
    screen_name: String,

    #[serde(rename = "type")]
    owner_type: OwnerType,

    #[serde(rename = "profileImageUrlLarge")]
    profile_image_url_large: Option<String>,

    #[serde(rename = "profileImageUrlMedium")]
    profile_image_url_medium: Option<String>,

    #[serde(rename = "profileImageUrlSmall")]
    profile_image_url_small: Option<String>,
}

#[derive(Serialize, Deserialize)]
pub struct Ratings {
    #[serde(rename = "rating")]
    rating: i64,
}

#[derive(Serialize, Deserialize)]
pub struct TableAuthor {
    #[serde(rename = "id")]
    id: Id,

    #[serde(rename = "displayName")]
    display_name: Name,

    #[serde(rename = "screenName")]
    screen_name: Name,

    #[serde(rename = "type")]
    table_author_type: OwnerType,
}

#[derive(Serialize, Deserialize)]
pub enum DisplayType {
    #[serde(rename = "fatrow")]
    Fatrow,

    #[serde(rename = "table")]
    Table,
}

#[derive(Serialize, Deserialize)]
pub enum ConsumerComplaintFlag {
    #[serde(rename = "default")]
    Default,

    #[serde(rename = "restorable")]
    Restorable,

    #[serde(rename = "restorePossibleForType")]
    RestorePossibleForType,
}

#[derive(Serialize, Deserialize)]
pub enum GrantFlag {
    #[serde(rename = "public")]
    Public,
}

#[derive(Serialize, Deserialize)]
pub enum GrantType {
    #[serde(rename = "viewer")]
    Viewer,
}

#[derive(Serialize, Deserialize)]
pub enum AvailableDisplayType {
    #[serde(rename = "fatrow")]
    Fatrow,

    #[serde(rename = "page")]
    Page,

    #[serde(rename = "table")]
    Table,
}

#[derive(Serialize, Deserialize)]
pub enum FieldType {
    #[serde(rename = "columnData")]
    ColumnData,

    #[serde(rename = "columnLabel")]
    ColumnLabel,
}

#[derive(Serialize, Deserialize)]
pub enum Width {
    #[serde(rename = "27%")]
    The27,

    #[serde(rename = "30%")]
    The30,

    #[serde(rename = "31%")]
    The31,

    #[serde(rename = "33%")]
    The33,

    #[serde(rename = "40%")]
    The40,
}

#[derive(Serialize, Deserialize)]
pub enum ModifyingViewUid {
    #[serde(rename = "s6ew-h6mp")]
    S6EwH6Mp,
}

#[derive(Serialize, Deserialize)]
pub enum OwnerType {
    #[serde(rename = "interactive")]
    Interactive,
}

#[derive(Serialize, Deserialize)]
pub enum Provenance {
    #[serde(rename = "official")]
    Official,
}

#[derive(Serialize, Deserialize)]
pub enum PublicationStage {
    #[serde(rename = "published")]
    Published,
}

#[derive(Serialize, Deserialize)]
pub enum Right {
    #[serde(rename = "read")]
    Read,
}

#[derive(Serialize, Deserialize)]
pub enum RowsUpdatedBy {
    #[serde(rename = "pjxg-ve4m")]
    PjxgVe4M,

    #[serde(rename = "54a3-qyun")]
    The54A3Qyun,

    #[serde(rename = "9e3m-2843")]
    The9E3M2843,

    #[serde(rename = "vvca-fr6g")]
    VvcaFr6G,
}

#[derive(Serialize, Deserialize)]
pub enum Name {
    #[serde(rename = "Dan Munz")]
    DanMunz,

    #[serde(rename = "Doug Taylor")]
    DougTaylor,

    #[serde(rename = "ming")]
    Ming,

    #[serde(rename = "Shannon Files")]
    ShannonFiles,
}

#[derive(Serialize, Deserialize)]
pub enum Id {
    #[serde(rename = "pjxg-ve4m")]
    PjxgVe4M,

    #[serde(rename = "54a3-qyun")]
    The54A3Qyun,

    #[serde(rename = "9e3m-2843")]
    The9E3M2843,

    #[serde(rename = "9eki-h2hn")]
    The9EkiH2Hn,
}

#[derive(Serialize, Deserialize)]
pub enum ViewType {
    #[serde(rename = "tabular")]
    Tabular,
}
